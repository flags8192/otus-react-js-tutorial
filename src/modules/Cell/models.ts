export enum ICellInfo {
  alive = 1,
  dead = 0,
}

export interface CellArray {
  size: number;
  data: ICellInfo[];
}

export type ISelectedCells = {
  [index: number]: boolean;
};
