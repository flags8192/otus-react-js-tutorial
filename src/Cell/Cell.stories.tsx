import React, { useState, FC } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Cell } from './Cell';
import { withKnobs, number, boolean } from '@storybook/addon-knobs';

export default {
  component: Cell,
  decorators: [withKnobs],
  title: 'Example/Cell',
} as ComponentMeta<typeof Cell>;

const Template: ComponentStory<typeof Cell> = (args) => <Cell {...args} />;

export const Static = Template.bind({});

Static.args = {
  num: number('num', 3),
  alive: boolean('alive', true),
  onClick: action('clicked'),
  isRight: boolean('isRight', true),
  isBottom: boolean('isBottom', true),
  isLeft: boolean('isLeft', true),
};

export const Dynamic: FC = () => {
  const [alive, setAlive] = useState(true);
  const num = number('num', 3);
  const onClick = () => {
    setAlive(!alive);
  };

  return (
    <Cell
      num={num}
      alive={alive}
      onClick={onClick}
      isRight={boolean('isRight', true)}
      isBottom={boolean('isBottom', true)}
      isLeft={boolean('isLeft', true)}
    />
  );
};
