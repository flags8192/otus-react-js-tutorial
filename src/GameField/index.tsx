export * from '@src/GameField/GameField';
