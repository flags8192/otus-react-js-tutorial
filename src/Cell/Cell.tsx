import React from 'react';
import styled from '@emotion/styled';
import {CELL_HEIGHT, CELL_WIDTH} from '@src/Cell/constatnts';

interface IContainer {
    isRight?: boolean;
    isLeft?: boolean;
    isBottom?: boolean;
    frameColor?: string;
    alive?: boolean;
}

const Container = styled.div<IContainer>`
  display: block;
  width: ${({isRight}) => CELL_WIDTH - (isRight ? 2 : 1)}px;
  height: ${({isBottom}) => CELL_HEIGHT - (isBottom ? 2 : 1)}px;
  text-align: center;
  border-top: ${({frameColor}) => '1px solid ' + frameColor};
  border-left: ${({frameColor}) => '1px solid ' + frameColor};
  border-right: ${({isRight, frameColor}) => '1px solid ' + isRight ? frameColor : ''}
  border-bottom: ${({isBottom, frameColor}) => '1px solid ' + isBottom ? frameColor : ''}
  cursor: pointer;
  float: left;
  ${({isLeft}) => (isLeft ? 'clear: both;' : '')}
  transition-property: background;
  transition-duration: 0.3s;
  transition-timing-function: ease-out;
  background: #555;
  background: ${({alive}) => (alive ? '#e44' : '#555')};
`;

interface IContent {
    alive?: boolean;
}

const Content = styled.span<IContent>`
  background: ${({alive}) => (alive ? '#e44' : '#555')};
`;

export interface CellProps {
    num: number;
    alive?: boolean;
    onClick: (num: number) => void;
    isLeft?: boolean;
    isRight?: boolean;
    isBottom?: boolean;
}

export const Cell: React.FC<CellProps> = ({
                                              num,
                                              alive = true,
                                              isLeft = true,
                                              isRight = true,
                                              isBottom = true,
                                              ...rest
                                          }) => {
    const handleClick = () => {
        rest.onClick(num);
    };
    const frameColor = '#ddd';

    return (
        <Container
            role="cellContainer"
            onClick={handleClick}
            isRight={isRight}
            isLeft={isLeft}
            alive={alive}
            frameColor={frameColor}
        >
            <Content role="cellContent" alive={alive}/>
        </Container>
    );
};
