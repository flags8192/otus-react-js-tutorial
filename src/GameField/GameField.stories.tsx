import React, { FC, useState, useEffect } from 'react';
import { withKnobs, number, boolean } from '@storybook/addon-knobs';
import { GameField } from '@src/GameField/index';
import { CellInfo } from '@src/Cell/models';
import { getInverted } from '@src/Cell/utils';

export default {
  component: GameField,
  decorators: [withKnobs],
  title: 'Example/GameField',
};

const START_WIDTH = 3;
const START_HEIGHT = 3;
const START_DATA: CellInfo[] = [];

export const Dynamic: FC = () => {
  const width = number('Width, cells', START_WIDTH);
  const height = number('Height, cells', START_HEIGHT);
  const showAll = boolean('showAll', false);

  const [data, setData] = useState(START_DATA);

  useEffect(() => {
    const cellsNumber = width * height;
    const startCellState = showAll ? CellInfo.alive : CellInfo.dead;
    const newData: CellInfo[] = [];
    for (let i = 0; i < cellsNumber; i++) {
      newData.push(startCellState);
    }
    setData(newData);
  }, [width, height, showAll]);

  const handleCellClick = (num: number) => {
    if (num < 0 || num >= data.length) {
      return;
    }

    const newData = data.concat();
    newData[num] = getInverted(newData[num]);
    setData(newData);
  };

  return <GameField data={data} onCellClick={handleCellClick} width={width} />;
};
