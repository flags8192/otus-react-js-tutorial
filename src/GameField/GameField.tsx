import React, { FC } from 'react';
import { CellInfo } from '@src/Cell/models';
import { Cell } from '@src/Cell';
import styled from '@emotion/styled';

export const FieldContainer = styled.section`
  background: #fff;
`;

export interface IGameField {
  data: CellInfo[];
  onCellClick: (num: number) => void;
  width: number;
}

export const GameField: FC<IGameField> = ({ onCellClick, width, data }) => {
  const widthMinus1 = width - 1;
  const heightMinus1 = Math.floor(data.length / width) - 1;
  return (
    <FieldContainer role="cellsGrid">
      {data.map((item, index) => {
        const y = Math.floor(index / width);
        const x = index % width;
        const isRight = x === widthMinus1;
        const isBottom = y === heightMinus1;
        const isLeft = x === 0;
        return (
          <Cell
            key={`${y}-${x}`}
            num={index}
            alive={item === CellInfo.alive}
            onClick={onCellClick}
            isLeft={isLeft}
            isRight={isRight}
            isBottom={isBottom}
          />
        );
      })}
      <div style={{ clear: 'both' }} />
    </FieldContainer>
  );
};
