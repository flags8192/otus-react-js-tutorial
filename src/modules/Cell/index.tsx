import React, { useContext } from 'react';
import styled from '@emotion/styled';
import { CELL_SIZE } from '@src/modules/Cell/constatnts';
import store from '@src/store';
import { setSelectedCell } from '@src/modules/Cell/actions';
import ICellInfo from '@src/model/ICellInfo';

interface IContainer {
  isRight?: boolean;
  isLeft?: boolean;
  isBottom?: boolean;
  frameColor?: string;
}
const Container = styled.div<IContainer>`
  display: block;
  width: ${({ isRight }) => CELL_SIZE - (isRight ? 2 : 1)}px;
  height: ${({ isBottom }) => CELL_SIZE - (isBottom ? 2 : 1)}px;
  text-align: center;
  border-top: ${({ frameColor }) => `1px solid ${frameColor}`};
  border-left: ${({ frameColor }) => `1px solid ${frameColor}`};
  border-right: ${({ isRight, frameColor }) => `1px solid ${isRight ? frameColor : ''};`}
  border-bottom: ${({ isBottom, frameColor }) => `1px solid ${isBottom ? frameColor : ''};`}
  float: left;
  ${({ isLeft }) => (isLeft ? 'clear: both;' : '')}
  transition-property: background;
  transition-duration: 0.3s;
  transition-timing-function: ease-out;
  background: #555;
  background: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;  
`;

interface IContent {
  selected?: boolean;
}
const Content = styled.div<IContent>`
  cursor: pointer;
  background: #ffff00;
  border-radius: 50%;
  background: ${({ selected }) => (selected ? '#ffff00' : '#ffff0070')};
  border-radius: 50%;
  width: 30px;
  height: 30px;
  margin: 5px;
  border: 0.6vh solid transparent;
  border-color: ${({ selected }) => (selected ? '#555' : '#9e9e9e')};

  :hover {
    background: #ffff00;
  }
`;

export interface CellProps {
  num: number;
  isFilled?: ICellInfo;
  isSelected?: boolean;
  isLeft?: boolean;
  isRight?: boolean;
  isBottom?: boolean;
}

const Cell: React.FC<CellProps> = ({
  num,
  isFilled = 0,
  isSelected = false,
  isLeft = true,
  isRight = true,
}) => {
  const { dispatch } = useContext(store);

  const handleClick = () => {
    dispatch(setSelectedCell(num, !isSelected));
  };
  const frameColor = '#ddd';

  return (
    <Container
      role={`cellContainer-${num}`}
      isRight={isRight}
      isLeft={isLeft}
      frameColor={frameColor}
    >
      {!!isFilled && (
        <Content role={`cellContent-${num}`} selected={isSelected} onClick={handleClick} />
      )}
    </Container>
  );
};

export default Cell;
