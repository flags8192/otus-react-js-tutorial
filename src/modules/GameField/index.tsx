import React, { FC, useContext } from 'react';
import styled from '@emotion/styled';

import { ICellInfo } from '@src/modules/Cell/models';
import Cell from '@src/modules/Cell';
import store from '@src/store';

export const FieldContainer = styled.section`
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  margin: auto;
  height: fit-content;
  width: fit-content;
`;

const GameField: FC = () => {
  const { state } = useContext(store);
  const { gameFieldData, gameFieldSize, selectedCells = {} } = state;
  const widthMinus1 = gameFieldSize - 1;
  const heightMinus1 = Math.floor(gameFieldData.length / gameFieldSize) - 1;

  if (!gameFieldSize) {
    return <p role="dataMessage">Нет данных!</p>;
  }

  return (
    <FieldContainer role="cellsGrid">
      {gameFieldData.map((item: ICellInfo, index: number) => {
        const y = Math.floor(index / gameFieldSize);
        const x = index % gameFieldSize;
        const isRight = x === widthMinus1;
        const isBottom = y === heightMinus1;
        const isLeft = x === 0;
        return (
          <Cell
            key={`${y}-${x}`}
            num={index}
            isFilled={item}
            isSelected={selectedCells[index]}
            isLeft={isLeft}
            isRight={isRight}
            isBottom={isBottom}
          />
        );
      })}
      <div style={{ clear: 'both' }} />
    </FieldContainer>
  );
};

export default GameField;
