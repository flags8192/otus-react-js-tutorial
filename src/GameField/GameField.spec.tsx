import React from 'react';
import { GameField } from './GameField';
import { render, screen, fireEvent } from '@testing-library/react';
import { CellInfo } from '@src/Cell/models';

let handleClick: (cellId: number) => void;
let data: CellInfo[];

describe('GameField', () => {
  beforeEach(() => {
    handleClick = jest.fn();
    data = [CellInfo.dead, CellInfo.dead, CellInfo.dead];
  });

  it('renders 3 cells from data:[3 items]', () => {
    const { asFragment } = render(<GameField data={data} onCellClick={handleClick} width={3} />);
    expect(asFragment()).toMatchSnapshot();
    const cellsGrid = screen.getByRole(/cellsGrid/gi);
    expect(cellsGrid).toBeInTheDocument();
    const cellContainer = screen.getAllByRole(/cellContainer/gi);
    expect(cellContainer.length).toBe(3);
  });

  it('calls onCellClick(<cell id>) callback when a cell is clicked', () => {
    render(<GameField data={data} onCellClick={handleClick} width={3} />);
    const cellContainer = screen.getAllByRole(/cellContainer/gi);
    fireEvent.click(cellContainer[1]);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
