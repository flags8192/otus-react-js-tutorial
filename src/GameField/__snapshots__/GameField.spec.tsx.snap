// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameField renders 3 cells from data:[3 items] 1`] = `
<DocumentFragment>
  <section
    class="css-1ce6rfj"
    role="cellsGrid"
  >
    <div
      class="css-1wsqmx8"
      role="cellContainer"
    >
      <span
        class="css-boqvn6"
        role="cellContent"
      />
    </div>
    <div
      class="css-alzgoi"
      role="cellContainer"
    >
      <span
        class="css-boqvn6"
        role="cellContent"
      />
    </div>
    <div
      class="css-1jsets4"
      role="cellContainer"
    >
      <span
        class="css-boqvn6"
        role="cellContent"
      />
    </div>
    <div
      style="clear: both;"
    />
  </section>
</DocumentFragment>
`;
